spanner/spannertest/db.go

```diff
@@ -705,38 +710,42 @@ func (t *table) alterColumn(alt spansql.AlterColumn) *status.Status {
 		return status.Newf(codes.InvalidArgument, "unknown column %q", alt.Name)
 	}
 
+	// TODO: check if the column isn't a primary key or array types.
+	t.cols[ci].NotNull = sct.NotNull
+
 	// Check and make type transformations.
 	oldT, newT := t.cols[ci].Type, sct.Type
 	stringOrBytes := func(bt spansql.TypeBase) bool { return bt == spansql.String || bt == spansql.Bytes }
 
-	// If the only change is adding NOT NULL, this is okay except for primary key columns and array types.
-	// We don't track whether commit timestamps are permitted on a per-column basis, so that's ignored.
-	// TODO: Do this when we track NOT NULL-ness.
-
-	// Change between STRING and BYTES is fine, as is increasing/decreasing the length limit.
-	// TODO: This should permit array conversions too.
-	if stringOrBytes(oldT.Base) && stringOrBytes(newT.Base) && !oldT.Array && !newT.Array {
-		// TODO: Validate data; length limit changes should be rejected if they'd lead to data loss, for instance.
-		var conv func(x interface{}) interface{}
-		if oldT.Base == spansql.Bytes && newT.Base == spansql.String {
-			conv = func(x interface{}) interface{} { return string(x.([]byte)) }
-		} else if oldT.Base == spansql.String && newT.Base == spansql.Bytes {
-			conv = func(x interface{}) interface{} { return []byte(x.(string)) }
-		}
-		if conv != nil {
-			for _, row := range t.rows {
-				if row[ci] != nil { // NULL stays as NULL.
-					row[ci] = conv(row[ci])
+	// TODO: We don't track whether commit timestamps are permitted on a per-column basis, so that's ignored.
+
+	if oldT != newT {
```

Comment:
It should be fine now as these are not pointers of a struct and have just primitive types. But, if there is a possibility that another field with non-primitive types would be added, we may want to do another comparing way like making a custom comparing func or using `reflect.DeepEqual`.

Reply:
yep, direct comparison of spansql.Type values is fine; it is done in dozens of places throughout this code.

Reply:
however, please invert this condition and early-out, and leave the remainder outdented like it was before.

Reply:
Ok, reverted the changes.
https://github.com/googleapis/google-cloud-go/pull/3491/commits/4df4f5db3958312d002897301242112326bb8239

I assume you intend to update here in another chance?

Reply:
Not sure what you mean by your last question, but all I was asking for was that when one writes something like

```go
if cond {
  ... // long section
  return x
}
return y
```

then it is often better to invert the condition to reduce the indentation of that long section:

```go
if !cond {
  return y
}
... // long section
return x
```

It makes it easier to follow what's going on.

Anyway, the current state of this PR looks fine now.

Reply:
Ok, I got what you mean. Thank you for your explanation and I agree with the basic principle you said.

Certainly, I could make the code like this.

```.go
if needsTypeRelatedTransformations { // means type transformation or length changes is demanded in the Alter Column statement
  if !stringOrBytes(oldT.Base) || !stringOrBytes(newT.Base) || oldT.Array && newT.Array{
   return y // error or kind of error
 }
  ... // long section
  return x // or go to another alternation process without noting
}
... // another alternation logics
return z // probably nil in this function.
```

And, I just wanted to confirm you are going to add the first IF statement above (and invert the condition as you say) next time because I just worried about a simple Statement adding/removing Not Null can't be accepted in the current code as long as it also has a type transformation.

But, as you say, it should be out of the scope of this PR and can fix anytime.

Thank you for your careful reviews 😄

Reply:
Ah, well if there's multiple transformations that need doing, the code needs more refactoring. Particularly if it needs to be able to fail (e.g. adding `NOT NULL` to a column with existing `NULL`s) then it likely can't handle one transformation completely at a time anyway, and would need to figure out the full change it'll make, check the data, then make the change all at once.

Reply:
Ok, I'm glad to know what you mean and let you know what I meant.

As you say, this function seems to need more refactoring. 
And, I guess there are several ways.
As you say, the code can figure out, check, and execute all demanding alternations at once.
Or, we can also do them per by a kind of alternation like Not Null alternation, type transformation and name change.

The previous change I tried to make is following the latter way, but I think either way fine.